lib/Basic/SignalHandlerWrapper.cpp

```diff
@@ -0,0 +1,35 @@
+//===- Support/SignalHandlerWrapper.cpp - Signal handling -----------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llbuild/Basic/SignalHandlerWrapper.h"
+
+#include <signal.h>
+
+using namespace llbuild;
+using namespace llbuild::basic;
+using namespace llbuild::basic::sys;
+
+void SignalHandlerWrapper::registerAction(int sig,
+                                          SignalHandlerWrapper *previous) {
+#if defined(_WIN32)
+  Handler handler = ::signal(sig, handler);
+  if (previous) {
+    previous->handler = handler;
+  }
+#else
+  struct sigaction action {};
+  action.sa_handler = handler;
+
+  struct sigaction previousHandler;
+  ::sigaction(sig, &action, &previousHandler);
+  if (previous) {
+    previous->handler = previousHandler.sa_handler;
```

Comment:
This isn't quite the same, because it won't restore the other parts of the old handler (like the mask or flags). I suggest making this an struct with a private void member which can then be casted in the implementation to a type containing the actual instance (which will have to be allocated, for simplicity) and that can then just have a `struct sigaction` inside.